### ISSUE TITLE
Investigate Whether Adding Emojis to Issue Labels Causes Conflicts with GitHub Actions

### ISSUE THREAD

**Issue by t-will-gillis** [hackforla/website#4050]
### Overview
We use issue labels to give a quick, visual summary of the issue contents. We want to explore whether adding emojis to text on our labels causes any unforeseen conflicts with existing GitHub Actions. 

### Details
- In the discussion during the Monday dev meeting (02-13-2023), one of the New Items mentioned the label system used by Hack for LA's Design Systems. These labels use emojis and a color-coding scheme to provide additional visual cues: [HfLA Design System Labels](https://github.com/hackforla/design-systems/labels). Notice especially the hourglass emoji and color scheme with the size label.

### Action Items
- [x] Determine which automations and actions might be affected by adding emojis to the labels. Examples include but are not limited to:
  - [x] [add-label.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/add-update-label-weekly/add-label.js)
  - [x] [sort-closed-issues.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/move-closed-issues/sort-closed-issues.js)
  - [x] [check-labels.js](https://github.com/hackforla/website/blob/gh-pages/github-actions/trigger-issue/add-missing-labels-to-issues/check-labels.js)
- [x] Research the effects of using emojis in labels to GitHub Actions, for example  `😴 2 weeks inactive`  and  `⏳ < 6 hrs`, to determine whether adding emojis would be an option or whether unexpected behavior results. 
- [x] Present findings to merge team.

### Resources/Instructions
[Hack for LA Design Systems label page](https://github.com/hackforla/design-systems/labels)
[HfLA Wiki: How to read and interpret labels](https://github.com/hackforla/website/wiki/How-to-read-and-interpret-labels)

**Comment by steven-positive-tran** [comment]
ETA: EOD 3/27
Availability: 7PM-12PM

**Comment by ExperimentsInHonesty** [comment]
@steven-positive-tran Please add your question to the issue since you have added it to the questions column

**Comment by steven-positive-tran** [comment]
I have a question on reviewing the effects of emoji. I tried doing making issues on my fork to see the github actions but it fails so I was thinking to just create new git actions just for my fork. Just wondering if you guys think theres a better way of doing this.

**Comment by t-will-gillis** [comment]
Hi @steven-positive-tran - this is something that you will be able to test on your fork. Here are some links that may be helpful:
https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions 
[Additional Notes for GitHub Actions_2023-3-25.pdf](https://github.com/hackforla/website/files/11070400/Additional.Notes.for.GitHub.Actions_2023-3-25.pdf)
[How to test GitHub Actions_2023-3-25.pdf](https://github.com/hackforla/website/files/11070401/How.to.test.GitHub.Actions_2023-3-25.pdf)

Please take a look at these. We know that GHAs can be difficult- if you would like we could figure out a time to discuss your progress and what might be holding you up. Are you able to make either the Tuesday or Thursday meetings? Feel free to Slack me with specific questions also.

**Comment by steven-positive-tran** [comment]
I apoloigize due to unforseen events I have to pause and have put this issue on the backlog.

**Comment by mjh-projects** [comment]
ETA: EOD 4/15/23
Availability: Everyweekday except Wednesday after 5:30 PM All day Saturday.

**Comment by mjh-projects** [comment]
Update
1. Progress - I have been reading on the actions in github to get acquainted
2. Blockers - Need more time to setup actions to expirement in my github
3. Availability Evenings after 6PM except Wednesday .
4. ETA - 4/28/2023

**Comment by mjh-projects** [comment]
Update
Progress
I had a project at work and was rather busy the past week so I have had to put off some of the research for this but will be starting it back u this week. 
 - I have been reading on the actions in github to get acquainted
Blockers - Need more time to setup actions to expirement in my github
Availability Evenings after 6PM except Wednesday .
ETA - 4/28/2023

**Comment by ExperimentsInHonesty** [comment]
@mjh-projects I am removing you from the issue.  We require weekly updates.  Please feel free to come back when you have the minimum amount of time required.  This issue should take no more than 6 hours, and thats the weekly requirement disclosed during onboarding.

**Comment by rdhmdhl** [comment]
availability: weekdays 5-9pm; weekends 9-5pm
ETA: 6/2

**Comment by rdhmdhl** [comment]
There are some issues with Github Actions that would be caused by adding emojis to labels. I've tested this by creating a new repository, copying the project board, and editing the workflow files to include my GitHub PAT token to have read/write permissions in my repository. 

**sort-closed-issues.js**
This action is used for checking labels on closed issues, and assigning them to the correct column on the project board. The HardLabel and SoftLabel move the issue into the "Done" column, while the OverrideLabel moves the issue into the "QA" column. If an emoji is added to any label used in this action, it won't be sorted properly on the project board unless the emoji is also added to this file.

A side note: HardLabels include visual changes like "Refactor CSS" or "Refactor HTML," and this GitHub action will move them to the "Done" column on the project board. The comment on line 42 mentions that "no visual change" is the reason for moving it to the "Done" column instead of the "QA" column.

**check-labels.js & post-labels-comment.js**
These actions are designed to add missing labels to issues and subsequently post a comment on the issue highlighting the labels that were added. However, there's a limitation when it comes to labels that include emojis. If an emoji is added to a label on Github, the current code will not recognize and add it to the issue without the emoji. This is due to the regular expression, which uses word boundaries, and assumes that labels are composed solely of alphanumeric characters or underscores, not considering emojis. 

This issue arises from this function:

```
function checkLabels(labels) {
  let labelsToAdd = []
  REQUIRED_LABELS.forEach((requiredLabel, i) => {
    const regExp = new RegExp(`\\b${requiredLabel}\\b`, 'gi')
    const isLabelPresent = labels.some(label => {
      // If the label is in the complexity exceptions array, it also fulfills the complexity requirements
      if (COMPLEXITY_EXCEPTIONS.includes(label) && requiredLabel === 'Complexity') {
        return true
      }
      return regExp.test(label)
    })
    if (isLabelPresent === false){
      labelsToAdd.push(LABEL_MISSING[i])
    }
  })
  return labelsToAdd
}
```
If we were to add emojis to labels in Github, this function would need to be adjusted to correctly identify labels containing emojis. Furthermore, it's important to note that while the code currently defaults to adding the word portion of the label if an emoji is included, any modifications must ensure that the entire label (including the emoji) is properly recognized and added.

**add-label.js**
This action adds a label based on the timeline of the last update made to an issue. The issue must be in the "In Progress" column for this action to run. The action uses a timeline of 3-14 days and evaluates the timeline for any changes to the issue. Since this is a long time to wait for testing purposes, I created a new timeline (Date.now()) and a new label to add after commenting on the issue.

Please note that this action will add any label that exists in the file, even if that label does not exist in the repository. If emojis are added to any label related to add-label.js, you'll need to update this file accordingly.

**Summary**
In summary, if a GHA contains a label variable that has a hard-coded string value, it will need to be updated to reflect the emoji that is added to a label in the repository. One potential solution to this is to create a file for pulling in the existing labels from Github every time a label changes, and then using a key/value pair for all of our Github actions utilizing labels. This would still require updating the checkLabels regular expression shown above. However, adding this will make the system more dynamic and adjust for changes to labels in the future.

**Comment by rdhmdhl** [comment]
Hi @t-will-gillis @roslynwythe @blulady ! Please see my comment above on the findings for this issue.

**Comment by t-will-gillis** [comment]
Hey @rdhmdhl thank you for all of your comments! 

Questions:

- When looking at an action such as `sort-closed-issues.js`: I believe you are saying that if the emoji was included both in the code and in labels themselves, the action should work with no issues, is this correct?
- Regarding the `function checkLabels(labels)` : I haven't tried matching emojis with regex yet- but doing a quick Google it looks like emojis might require an extra step or two but would not automatically preclude regex matching, what are your thoughts?
- Finally- the big question: given that the emoji can be added to the label itself, did you find any cases where the action would break if the reference in the js file/ action matched the emoji-label?

**Comment by rdhmdhl** [comment]
hey @t-will-gillis!

- Correct, _if_ the emoji is included in the code and the label in Github, `sort-closed-issues.js` will work with no issues. 

- From my research and understanding, emojis are unicode, and the word boundary (`\b`) in regular expressions has limitations when it comes to unicode characters beyond ASCII. Therefore, an emoji at the start or end of a string might not be matched correctly using `\b`.

The current implementation uses `'\\b${myLabel}\\b'` where `\\b` is designed to match the exact word contained in the 'myLabel" variable. It will only match occurrences of "myLabel" that appear as a whole word, rather than part of a larger word. I'm not great at regular expressions, but a solution might look something like this:

`const regExp = new RegExp(`(^|\\s)${myLabel}(?=\\s|$)`, 'gi');`

(^|\\s) matches the start of the line or a space.
${myLabel} matches the variable content.
(?=\\s|$) is a lookahead that checks for a space or the end of the line following myLabel. It doesn't consume characters in the string, meaning the character(s) that follow myLabel are not part of the match and can be matched again.

I'd recommend running more tests before implementing. 

- In the files I reviewed, I didn't find anything that would throw an error, but doing so would cause unexpected behavior. In the check-labels.js Github action, specifically for the checkLabels regex function. I've searched the repo for more regex functions and was not able to find any more pertaining to labels.

**Comment by rdhmdhl** [comment]
1. I am writing a simplified version of my findings, and I'll post it here when completed
2. It was difficult to setup a test for this and simulate the environment where and when these actions run
3. I will have 8 hours this week to work on this, but it likely will not take that long to finish
4. 6/13

**Comment by t-will-gillis** [comment]
Update- @rdhmdhl and I have discussed his findings and he indicated that he plans to post his summary shortly.

**Comment by rdhmdhl** [comment]
Adding emojis to labels in GitHub Actions is doable and won't require extensive refactoring. If we decide to use emojis, we need to follow these steps to implement them successfully:

- Update the code in the affected actions, such as sort-closed-issues.js, check-labels.js, and post-labels-comment.js, to recognize and handle labels with emojis.
- Ensure that the labels used in the code match the labels in the repository, including any emojis added to them.
- Consider creating a file that contains all existing labels from GitHub and uses a key/value pair approach for all actions utilizing labels. These key value pairs would look something like -- `"roleDesign": "role: design"`, or `"roleDesign": "role: design 🎨"`. This will make the system more maintainable for future label changes.

**Comment by t-will-gillis** [comment]
@ExperimentsInHonesty above is @rdhmdhl 's summary for adding emojis to issues, for consideration and next steps.

**Comment by rdhmdhl** [comment]
Reviewed findings on 7/9 with the Team, Bonnie was not present. 

Was told to join the meeting on 7/10, or 7/16 to review with Bonnie and close this issue.

**Comment by ExperimentsInHonesty** [comment]
When you finish writing up the decision record
- [ ] We are currently in the process of moving the old wiki to the new [website-wiki repo](https://github.com/hackforla/website-wiki), so we will not be making any changes or additions to the old wiki at this time. Thus, we will be adding wiki content through a different process now. Read [How to Contribute to the Wiki](https://github.com/hackforla/website/wiki/How-to-Contribute-to-the-Wiki)

**Comment by rdhmdhl** [comment]
Will have the decision record wrote by 7/21 EOD.

**Comment by rdhmdhl** [comment]
### DR: Adding emojis to labels, updating Github Actions

Issue
https://github.com/hackforla/website/issues/4050

Problem Statement
Labels are hard-coded and spread throughout our Github Action files. If a label is changed in the repo, we have to investigate where that label was being used in Github Actions, and update it. 

This issue explores the implications of adding emojis to our Github Action files, and if it would cause any unforeseen errors. I tested this issue by setting up my own testing environment, copying the project board from HFLA, and pasting it into my own repository. I then used a personal access token for the GHAs to access the project issues, project board and pull requests. 

Potential Solution
Use a central file containing all labels, in a key/value pair formatting. Import this file to any Github Action that uses labels, and reference the key that contains the correct label. 

Feasibility Determination
Adding emojis to labels in GitHub Actions is doable and won't require extensive refactoring. If we decide to use emojis, we need to follow these steps to implement them successfully:

- Update the code in the affected actions, such as sort-closed-issues.js, check-labels.js, and post-labels-comment.js, to recognize and handle labels with emojis.
- Create a file that contains all existing labels from GitHub and uses a key/value pair approach for all actions utilizing labels. These key value pairs would look something like -- "roleDesign": "role: design", or "roleDesign": "role: design 🎨". 
- Import this file to each GHA that uses a label, and reference the key for the desired label. This will make the system more maintainable for future label changes.

Plan Summary:
- Add decision record
- Create a large issue for adding in the central file for all labels, create medium issues to update a GHA that contain labels. Test all labels/GHAs that were updated.

**Comment by roslynwythe** [comment]
Thank you @rdhmdhl for a thorough analysis and clear write-up.  I really like your idea for the key-value pair file for labels, as it will help ease the rollout of any label changes, not just for emojis.  

I just have a few questions.  

1) It seems obvious now but could you add a sentence of introduction in the Problem Statement of the DR, just stating the need to explore whether adding emojis to text on our labels causes any unforeseen conflicts with existing GitHub Actions.
2) Are there any other HfLA actions aside from those you have discussed (add-label.js, sort-closed-issues.js and check-label.js) that might be affected by adding emojis to labels?
3) Please advise, would it be feasible to split up the large issue that you propose as follows:
Issue 1: Create the key-value pair file and incorporate its use into sort-closed-issues.js
Issue 2: Incorporate key-value pair into check-labels.js and update regex
Then additional issues, each to incorporate the key-value pair into one action script.

**Comment by rdhmdhl** [comment]
Hi @roslynwythe! I've update my previous comment to include an introduction of the issue. 

Please see answers below:

**Actions that will not be affected by adding emojis to labels**
_create-instructions.js
post-comment.js
preliminary-update-comment.js
get-project-data.js
comment-issue.js
get-list.js
contributors-data.js
find-linked-issue.js_


**Actions that do use labels:**
_sort-closed-issues.js
add-label.js
check-labels.js
post-labels-comments.js
create-new-issue.js
obtain-labels.js_ -- gets the labels from an issue, returns them. 
_check-label-preliminary-update.js_ -- this checks the current labels on an issue. 
_set-pr-labels.js_ -- this file will post labels on a PR
`const res = await _setLabels({ github, context, core }, pr, labels);`


These could potentially be split up into many different medium or large issues, and have the HFLA member adjust each file that uses labels as hardcoded values.

**Comment by roslynwythe** [comment]
Thank you @rdhmdhl for excellent work on this DR.  It has been approved by dev leads and PM.  I added it to the [Decision Records on Solutions Adopted](https://github.com/hackforla/website/wiki/Decision-Records-on-Solutions-Adopted) page and will close this issue as completed.  As discussed above, we would like to proceed by creating an issue for development of the key-value pair file and integrating that into a singe script such as sort-closed-issues.js (or other if that would be make more sense).  Then we can manage the others.  It would be great if you could write that first issue.  Please let me know if that is something you are interested in doing.

**Comment by rdhmdhl** [comment]
Thank you @roslynwythe! I will absolutely write the first issue & help out in any other areas regarding GHAs.
